cdc/sinkv2/eventsink/mq/producer/kafka_producer.go

```diff
@@ -0,0 +1,240 @@
+	close(k.closedChan)
+	k.closed = true
+	// `client` is mainly used by `asyncProducer` to fetch metadata and other related
+	// operations. When we close the `kafkaSaramaProducer`, TiCDC no need to make sure
+	// that buffered messages flushed.
+	// Consider the situation that the broker does not respond, If the client is not
+	// closed, `asyncProducer.Close()` would waste a mount of time to try flush all messages.
+	// To prevent the scenario mentioned above, close client first.
+	start := time.Now()
+	if err := k.client.Close(); err != nil {
+		log.Error("close sarama client with error", zap.Error(err),
+			zap.Duration("duration", time.Since(start)),
+			zap.String("namespace", k.id.Namespace),
+			zap.String("changefeed", k.id.ID), zap.Any("role", k.role))
+		return err
+	}
+	log.Info("sarama client closed", zap.Duration("duration", time.Since(start)),
+		zap.String("namespace", k.id.Namespace),
+		zap.String("changefeed", k.id.ID), zap.Any("role", k.role))
+
+	start = time.Now()
+	err := k.asyncProducer.Close()
+	if err != nil {
+		log.Error("close async client with error", zap.Error(err),
+			zap.Duration("duration", time.Since(start)),
+			zap.String("namespace", k.id.Namespace),
+			zap.String("changefeed", k.id.ID),
+			zap.Any("role", k.role))
+		return err
+	}
+	log.Info("async client closed", zap.Duration("duration", time.Since(start)),
+		zap.String("namespace", k.id.Namespace),
+		zap.String("changefeed", k.id.ID), zap.Any("role", k.role))
+
+	return nil
+}
+
+func (k *kafkaProducer) run(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-k.closedChan:
+			return nil
+		case err := <-k.failpointCh:
+			log.Warn("receive from failpoint chan", zap.Error(err),
+				zap.String("namespace", k.id.Namespace),
+				zap.String("changefeed", k.id.ID), zap.Any("role", k.role))
+			return err
+		case ack := <-k.asyncProducer.Successes():
+			if ack != nil {
+				callback := ack.Metadata.(messageMetaData).callback
+				if callback != nil {
+					callback()
+				}
+			}
+		case err := <-k.asyncProducer.Errors():
+			// We should not wrap a nil pointer if the pointer
+			// is of a subtype of `error` because Go would store the type info
+			// and the resulted `error` variable would not be nil,
+			// which will cause the pkg/error library to malfunction.
+			if err == nil {
+				return nil
+			}
+			return cerror.WrapError(cerror.ErrKafkaAsyncSendMessage, err)
+		}
+	}
+}
+
+// NewKafkaProducer creates a new kafka producer.
+func NewKafkaProducer(
+	ctx context.Context,
+	client sarama.Client,
+	errCh chan error,
+) (Producer, error) {
+	changefeedID := contextutil.ChangefeedIDFromCtx(ctx)
+	role := contextutil.RoleFromCtx(ctx)
+	log.Info("Starting kafka sarama producer ...",
+		zap.String("namespace", changefeedID.Namespace),
+		zap.String("changefeed", changefeedID.ID), zap.Any("role", role))
+
+	asyncProducer, err := sarama.NewAsyncProducerFromClient(client)
+	if err != nil {
```

Comment:
we can just:
`return nil, cerror.WrapError(cerror.ErrKafkaNewSaramaProducer, err)`

Reply:
I am wrong , nerver mind.